Add Offset for PF

pap1723 opened the pull request.

DRVeyl commented on Source/ProceduralFairings/FairingSide.cs:
```diff
@@ -842,5 +843,26 @@ public void rebuildMesh ()
 
             m.triangles = tri;
         }
+
+        /*
+        public void SetOffset(Vector3 offset)
+        {
+            var mf = part.FindModelComponent<MeshFilter>("model");
+            mf.transform.localPosition = offset;
+        }
+        */
```
Just delete this rather than leaving commented code.

DRVeyl commented on Source/ProceduralFairings/FairingSide.cs:
```diff
@@ -842,5 +843,26 @@ public void rebuildMesh ()
+            mf.transform.localPosition = offset;
+        }
+        */
+        public IEnumerator SetOffset(Vector3 offset, float time)
```
Worth defaulting `time` to something?  eg 0.3?

DRVeyl commented on Source/ProceduralFairings/FairingBase.cs:
```diff
@@ -117,6 +121,13 @@ public enum BaseMode { Payload, Adapter, Plate, Other }
         public ModuleDecouple Decoupler;
+        private readonly List<ProceduralFairingSide> sides = new List<ProceduralFairingSide>();
+        public Vector3 editorOpenOffset => new Vector3(OffsetValueX, 0, 0);
+        public float openSpeed = 0.3f;
```
Is this `const`?

DRVeyl commented on Source/ProceduralFairings/FairingBase.cs:
```diff
@@ -107,6 +107,10 @@ public enum BaseMode { Payload, Adapter, Plate, Other }
         [UI_Toggle(disabledText = "Off", enabledText = "On")]
         public bool showInterstageNodes = true;
 
+        [UI_Toggle(scene = UI_Scene.Editor, disabledText = "Closed", enabledText = "Open", suppressEditorShipModified = false)]
```
suppressEditorShipModified = false is the default, doesn't need to be specified

DRVeyl commented on Source/ProceduralFairings/FairingBase.cs:
```diff
@@ -346,6 +357,9 @@ void SetUIChangedCallBacks()
 
             Fields[nameof(showInterstageNodes)].uiControlEditor.onFieldChanged += OnNodeVisibilityChanged;
             Fields[nameof(showInterstageNodes)].uiControlEditor.onSymmetryFieldChanged += OnNodeVisibilityChanged;
+            
+            //Fields[nameof(openFairing)].uiControlEditor.onFieldChanged += OnToggleOpen;
+            //Fields[nameof(openFairing)].uiControlEditor.onSymmetryFieldChanged += OnToggleOpen;
```
Since you're relying on OnEditorShipModified to fire to start redrawing things, and not using these, just remove them.

DRVeyl commented on Source/ProceduralFairings/FairingBase.cs:
```diff
@@ -1376,6 +1394,8 @@ public void recalcShape ()
                         float off = sn.position.y;
                         off -= oppNodePos.y;
                         sf2.meshPos = new Vector3(-norm.magnitude, -off, 0);
+                        var currentOffset = sf2.meshPos + (openFairing ? editorOpenOffset : Vector3.zero);
+                        StartCoroutine(sf2.SetOffset(currentOffset, openSpeed));
```
Use the nice UpdateOpen() method you made?  Or get rid of it if nothing will call it?

pap1723 committed:
Update gitignore to handle Rider

pap1723 committed:
Add Offset Ability to PF

pap1723 commented:
Updated

DRVeyl commented on Source/ProceduralFairings/FairingSide.cs:
```diff
@@ -842,5 +843,19 @@ public void rebuildMesh ()
+            var lp = mf.transform.localPosition;
+            float elapsedTime = 0f;
+
+            while (elapsedTime < time)
+            {
+                mf.transform.localPosition = Vector3.Lerp(lp, offset, (elapsedTime / time));
+                elapsedTime += Time.deltaTime;
+                yield return null;
+            }
```
This will stop just short of offset, because the last loop execution is for elapsedTime < time.
Upon exiting the while, add one last line:  `mf.transform.localPosition = offset`

pap1723 committed:
Change SetOffset to check for <= in order to make the last movement

pap1723 committed:
ACTUALLY fixed the FairingSide Offset

pap1723 committed:
Add the Handlers again to Stop offset from "bouncing" on craft changes

This was due to the coroutine. Now the coroutine is only called from the actual PAW button.

DRVeyl commented on Source/ProceduralFairings/FairingBase.cs:
```diff
@@ -117,6 +121,13 @@ public enum BaseMode { Payload, Adapter, Plate, Other }
         readonly List<ConfigurableJoint> joints = new List<ConfigurableJoint>();
         public DragCubeUpdater dragCubeUpdater;
         public ModuleDecouple Decoupler;
+        private readonly List<ProceduralFairingSide> sides = new List<ProceduralFairingSide>();
```
This is no longer used I believe, and can be removed.

DRVeyl committed:
Support Open/Close Toggle in Editor on FairingSide

Toggle the fairingbase open/closed setting.
Do so by finding the UIControl and invoking it the same way the PAW does.

DRVeyl committed:
Scale fairing open offset with fairingbase size
